Migrate data structures

alxmjo opened the pull request:
- Migrate singly linked list
- Migrate binary search tree

@faheel Besides the fixes you pointed out today and a few other little things here and there (migrating `utils.hpp` to the proper `sorting` folder, for one), I think this is the final step of converting all existing code into the new format. Only took us two years!  😄

I'll leave this open until you've had a chance to look it over.

alxmjo committed:
Migrate singly linked list

- Create required directories for implementation and tests
- Move files into proper directories
- Migrate singly_linked_list.cpp to header-only implementation
- Add unit tests for singly_linked_list.hpp
- Update README
- Update build script

alxmjo committed:
Migrate binary search tree

- Create required directories for implementation and tests
- Move files into proper directories
- Migrate binary_search_tree.cpp to header-only implementation
- Add unit tests for binary_search_tree.hpp
- Update README
- Update build script

alxmjo commented:
The clang++5.0 build failed with the following error code:

```
The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends $(travis_apt_get_options) install clang-5.0" failed and exited with 100 during .
```

Can't determine whether this is my problem or not. More info [here](https://travis-ci.community/t/build-fails-on-ubuntu-14-04-with-error-the-command-sudo-e-apt-get-yq-no-install-suggests-no-install-recommends-travis-apt-get-options-install-powershell-failed-and-exited-with-100-during/3116).

alxmjo commented:
I just re-ran the build associated with #250 without issue, which makes me think that there's something up with this PR. On the other hand, [this](https://travis-ci.community/t/cannot-apt-get-install-clang-5-0/3250) seems to indicate an issue with using Trusty as the default build environment. Apparently Travis is migrating to Xenial.

alxmjo committed:
Attempted travis fix

alxmjo committed:
Attempted travis fix

alxmjo commented:
The most recent commit appears to have done the trick. Specified Trusty manually, though it seems like at some point we'll want to [migrate this to Xenial](https://changelog.travis-ci.com/xenial-as-the-default-build-environment-99476).

faheel commented on .travis.yml:
```diff
@@ -42,6 +42,7 @@ matrix:
     # Linux Clang 5 build
     - os: linux
       compiler: clang
+      dist: trusty
```
I think removing `dist` and `sources` should work as Clang 5.0 (along with 4.0 and 6.0) is available in the default Xenial repos (from the [link you shared](https://travis-ci.community/t/cannot-apt-get-install-clang-5-0/3250/3)).

For now, I suggest dropping the commits to this file and letting the build fail (since we know that the tests run fine otherwise) and then later updating the Travis CI config in a separate PR where we can also target GCC 8, Clang 6 and the latest macOS image.

alxmjo replied:
Reverted to the original format.

faheel commented on C++/README.md:
```diff
@@ -46,9 +46,9 @@ Implementation of well-known (and some rare) algorithms, in C++.
-* [Singly linked list](include/data_structures/singly_linked_list.cpp)
+* [Binary search tree](include/data_structure/tree/binary_search_tree.hpp) :white_check_mark:
+* [Disjoint-set](include/data_structure/set/disjoint_set.hpp) :white_check_mark:
+* [Singly linked list](include/data_structure/linked_list/singly_linked_list.hpp) :white_check_mark:
```
Would be nice to arrange these under their respective topic's list, i.e. trees, sets and linked lists.

alxmjo replied:
Done.

faheel commented on C++/include/data_structure/linked_list/singly_linked_list.hpp:
```diff
@@ -0,0 +1,394 @@
+	Node<T> *next;
+	Node(void);
+
+public:
```
This is not required as `struct` methods and members are public by default.

alxmjo replied:
I think you're right that this was originally meant to be a class, so the `public` modifier makes more sense. With that in mind I've switched this to class and left in the modifier.

faheel commented on C++/include/data_structure/linked_list/singly_linked_list.hpp:
```diff
@@ -0,0 +1,394 @@
+	T& get_value(void);
+	void set_value(const T&);
+	Node<T> *get_next(void);
+	void set_next(Node<T> * const);
```
Getters and setters are usually not used for structures. Or is this supposed to be a class, in which case the `struct` declaration has been mistakenly used instead of `class`?

alxmjo replied:
Switched to class.

faheel commented on C++/include/data_structure/linked_list/singly_linked_list.hpp:
```diff
@@ -0,0 +1,394 @@
+public:
+	Node(const T&, Node<T> * const);
+	Node(const Node&);
+	Node<T>& operator=(const Node<T> &);
```
In the parameter type, `&` should not have a space before it.

faheel reviewed (changes requested):
At a few places, some function parameters and local variables are written as:
```cpp
Node<T> * const some_var
Node<T> & other_var
```
which are not wrong, but I think it would be better to write these as:
```cpp
const Node<T>* some_var
Node<T>& other_var
```
since its easier to determine the type of a variable (`const Node<T>*` and `Node<T>&`) this way.

alxmjo committed:
Implement requested changes

- Convert Node from struct to class
- Standardize Node<T> reference and pointer placement
- Update README to add categories for data structures
- Revert build script (so that changes can be made in separate PR)

alxmjo commented:
Requested changes submitted, including the issues with the Node<T> references and pointers.

faheel commented on C++/test/data_structure/tree/binary_search_tree.cpp:
```diff
@@ -0,0 +1,40 @@
+	REQUIRE(tree.search(1) == true);
+	tree.remove(1);
+	REQUIRE(tree.search(1) == false);
+	REQUIRE(tree.search(2) == true);
```
Having more insertions, removals and searches here would be good as that would make the test more robust, and therefore, more reliable.

alxmjo replied:
Good thing you asked for this, because it turns out that removing three elements in a row throws a SIGABRT. I'll run that down and add it to a new commit.

alxmjo replied:
Looks like `remove()` was missing a return statement. Added a bunch of tests and they're all passing. Thanks again for the nudge.

faheel commented on C++/test/data_structure/linked_list/singly_linked_list.cpp:
```diff
@@ -0,0 +1,46 @@
+TEST_CASE("Add elements", "[singly-linked-list]") {
+	SinglyLinkedList<int> list;
+	list.insert_rear(2);
+	list.insert_rear(3);
```
None of the tests cases singly linked list make use of `insert_front()`, `operator[]` or `is_empty()`. Calling these functions in various unit tests will improve code coverage.

alxmjo committed:
Implement unit test changes

- Add tests for insert_front()
- Add tests for operator[]
- Add test for is_empty()
- Expand test cases

alxmjo committed:
Implement unit test changes

- Expand unit tests
- Add missing return statement
- Fix typos
- Add braces

alxmjo commented:
Changes and fixes implemented. Ready to merge as far as I'm concerned.

alxmjo committed:
Clean up white space

faheel commented:
Nice work @alxmjo!
